src/v/cluster/controller.cc

```diff
@@ -217,551 +219,499 @@ ss::future<> controller::start(
+                "({}) and actual log size ({}).  If the storage "
+                "directory was not modified intentionally, this is a "
+                "bug.",
+                last_applied,
+                disk_dirty_offset);
+
+              // Try waiting for replay to reach the disk_dirty_offset,
+              // ignore last_applied.
+              return stm.wait(disk_dirty_offset, model::time_from_now(5s))
+                .handle_exception_type([](const ss::timed_out_error&) {
+                    // Ignore timeout: it just means the controller
+                    // log replay is done without hitting the disk
+                    // log hwm (truncation happened), or that we were
+                    // slow and controller replay will continue in
+                    // the background while the rest of redpanda
+                    // starts up.
+                });
           }
-          cloud_storage_clients::bucket_name bucket = bucket_opt.value();
-          _metadata_uploader = std::make_unique<cloud_metadata::uploader>(
-            _raft_manager.local(),
-            _storage.local(),
-            bucket,
-            _cloud_storage_api.local(),
-            _raft0,
-            _tp_state.local(),
-            offsets_uploader);
-          if (config::shard_local_cfg().enable_cluster_metadata_upload_loop()) {
-              _metadata_uploader->start();
+
+          vlog(
+            clusterlog.info,
+            "Controller log replay starting (to offset {})",
+            last_applied);
+
+          if (last_applied == model::offset{}) {
+              return ss::now();
+          } else {
+              // The abort source we use here is specific to our startup
+              // phase, where we can't yet safely use our member abort
+              // source.
+              return stm.wait(last_applied, model::no_timeout, as);
           }
-          _recovery_backend
-            = std::make_unique<cloud_metadata::cluster_recovery_backend>(
-              _recovery_manager.local(),
+      });
+    vlog(clusterlog.info, "Controller log replay complete.");
+
+    /// Once the controller log is replayed and topics are recovered
+    /// print the RF minimum warning
+    _tp_frontend.local().print_rf_warning_message();
+
+    co_await cluster_creation_hook(discovery);
+
+    // start shard_balancer before controller_backend so that it boostraps
+    // shard_placement_table and controller_backend can start with already
+    // initialized table.
+    co_await _shard_balancer.invoke_on(
+      shard_balancer::shard_id,
+      &shard_balancer::start,
+      conf_invariants.core_count);
+
+    if (conf_invariants.core_count > ss::smp::count) {
+        // Successfully starting shard_balancer with reduced core count means
+        // that all partition info from extra kvstores has been copied and we
+        // can finally update the configuration invariants.
+        co_await _storage.local().kvs().put(
```

Comment:
nit: maybe we can log those events with `INFO` level to make it easy to understand when this happens